File: frontend/src/features/editor/trackStack.js
// Pila de capas de la timeline (como CapCut).
//
// Las pistas de VÍDEO y de TEXTO forman UNA sola pila, en el orden del array
// `tracks`: índice 0 = fondo, la última = delante. El audio no pinta nada, así
// que queda fuera de la pila. Preview y export componen por este orden, así que
// un texto puede quedar DETRÁS de un vídeo (texto detrás de una persona).
//
// En pantalla la pila va invertida (la de delante arriba) y el audio debajo, en
// su orden de array.

// Versión del formato del timeline que habla el editor (= CURRENT_SCHEMA_VERSION
// de backend/app/migrations.py). Va en todo lo que se envía: sin ella el backend
// lo trata como legado y lo migra, y la migración v4→v5 volvería a subir los
// textos por encima del vídeo, deshaciendo el orden de capas del usuario.
export const TIMELINE_SCHEMA_VERSION = 5

export const isStackTrack = (t) => t?.kind === 'video' || t?.kind === 'text'

export function displayTracks(tracks) {
  const list = tracks || []
  return [...list.filter(isStackTrack).reverse(), ...list.filter((t) => !isStackTrack(t))]
}

/** track_id → capa (0 = fondo). Las pistas de audio no aparecen. */
export function stackLayers(tracks) {
  const layers = new Map()
  let i = 0
  for (const t of tracks || []) if (isStackTrack(t)) layers.set(t.id, i++)
  return layers
}

/**
 * Índice del array para una pista nueva de `kind` sin cambiar lo que se ve:
 * el vídeo va encima del vídeo más alto (los textos que hubiera delante siguen
 * delante); texto y audio al final (el texto queda arriba del todo de la pila:
 * el audio no cuenta en ella).
 */
export function defaultTrackIndex(tracks, kind) {
  const list = tracks || []
  if (kind === 'video') return list.findLastIndex((t) => t.kind === 'video') + 1
  return list.length
}

export function insertTrack(tracks, track, index) {
  const list = [...(tracks || [])]
  const at = Number.isInteger(index) ? Math.max(0, Math.min(list.length, index)) : defaultTrackIndex(list, track.kind)
  list.splice(at, 0, track)
  return list
}

// Pantalla → array: la pila vuelve a su orden (fondo primero) y el audio detrás.
function fromDisplay(rows) {
  return [...rows.filter(isStackTrack).reverse(), ...rows.filter((t) => !isStackTrack(t))]
}

/**
 * Coloca `trackId` justo encima (`above`) o debajo (`below`) de `targetId` tal
 * como se ven en pantalla. Solo dentro de su grupo: la pila (vídeo + texto) o el
 * audio. Devuelve el mismo array si no hay nada que mover.
 */
export function reorderTrack(tracks, trackId, targetId, place) {
  const rows = displayTracks(tracks)
  const moving = rows.find((t) => t.id === trackId)
  const target = rows.find((t) => t.id === targetId)
  if (!moving || !target || moving === target || isStackTrack(moving) !== isStackTrack(target)) return tracks
  const rest = rows.filter((t) => t !== moving)
  rest.splice(rest.indexOf(target) + (place === 'below' ? 1 : 0), 0, moving)
  const next = fromDisplay(rest)
  return next.every((t, i) => t === tracks[i]) ? tracks : next
}

/** Vecina en pantalla dentro del grupo: dir -1 = la de arriba, 1 = la de abajo. */
export function trackNeighbor(tracks, trackId, dir) {
  const rows = displayTracks(tracks)
  const i = rows.findIndex((t) => t.id === trackId)
  const n = rows[i + dir]
  return i >= 0 && n && isStackTrack(n) === isStackTrack(rows[i]) ? n : null
}

/** Sube (dir -1) o baja (dir 1) una pista un puesto en pantalla. */
export function stepTrack(tracks, trackId, dir) {
  const n = trackNeighbor(tracks, trackId, dir)
  return n ? reorderTrack(tracks, trackId, n.id, dir < 0 ? 'above' : 'below') : tracks
}

File: backend/app/track_stack.py
"""Pila de capas de la timeline (espejo de ``frontend/.../editor/trackStack.js``).

Las pistas de VÍDEO y de TEXTO forman UNA sola pila, en el orden de
``timeline.tracks``: índice 0 = fondo, la última = delante. El audio no pinta
nada y queda fuera. Preview y export componen por este orden, así que un texto
puede quedar DETRÁS de un vídeo (texto detrás de una persona recortada).

En pantalla la pila va invertida (la de delante arriba) y el audio debajo.
"""
from __future__ import annotations

STACK_KINDS = ("video", "text")


def is_stack_track(t) -> bool:
    return getattr(t, "kind", None) in STACK_KINDS


def stack_layers(tracks) -> dict[str, int]:
    """track_id → capa (0 = fondo). Las pistas de audio no aparecen."""
    return {t.id: i for i, t in enumerate(t for t in tracks if is_stack_track(t))}


def default_track_index(tracks, kind: str) -> int:
    """Índice para una pista nueva sin cambiar lo que se ve: el vídeo encima del
    vídeo más alto (los textos de delante siguen delante); texto y audio al final
    (el texto queda arriba del todo de la pila: el audio no cuenta en ella)."""
    if kind == "video":
        return max((i for i, t in enumerate(tracks) if t.kind == "video"), default=-1) + 1
    return len(tracks)


def insert_track(tracks: list, track, index: int | None = None) -> None:
    """Inserta ``track`` en ``tracks`` (in place): en ``index`` o en su sitio por defecto."""
    at = default_track_index(tracks, track.kind) if index is None else max(0, min(len(tracks), index))
    tracks.insert(at, track)


def display_order(tracks) -> list:
    stack = [t for t in tracks if is_stack_track(t)]
    return stack[::-1] + [t for t in tracks if not is_stack_track(t)]


def reorder_track(tracks, track_id: str, target_id: str, place: str) -> list:
    """Nueva lista con ``track_id`` justo encima (``above``) o debajo (``below``)
    de ``target_id`` tal como se ven en pantalla, dentro de su grupo (pila o audio)."""
    if place not in ("above", "below"):
        raise ValueError("place debe ser 'above' o 'below'")
    rows = display_order(tracks)
    moving = next((t for t in rows if t.id == track_id), None)
    target = next((t for t in rows if t.id == target_id), None)
    if moving is None or target is None:
        raise ValueError(f"pista desconocida: {track_id if moving is None else target_id}")
    if moving is target:
        return list(tracks)
    if is_stack_track(moving) != is_stack_track(target):
        raise ValueError("una pista de audio solo se reordena entre pistas de audio "
                         "y una de vídeo/texto dentro de la pila de vídeo/texto")
    rest = [t for t in rows if t is not moving]
    rest.insert(rest.index(target) + (1 if place == "below" else 0), moving)
    stack = [t for t in rest if is_stack_track(t)]
    return stack[::-1] + [t for t in rest if not is_stack_track(t)]


def buried_text_track_ids(tracks) -> set[str]:
    """Pistas de texto con alguna pista de vídeo por encima en la pila. Sus textos
    se componen en su capa, no en el .ass global (que va encima de todo)."""
    buried: set[str] = set()
    video_above = False
    for t in reversed([t for t in tracks if is_stack_track(t)]):
        if t.kind == "video":
            video_above = True
        elif video_above:
            buried.add(t.id)
    return buried
